qiskit_nature/operators/second_quantization/spin_op.py

```diff
@@ -437,35 +450,30 @@ def to_matrix(self) -> np.ndarray:
         return mat
 
     def _from_sparse_label(self, labels):
-        num_terms = len(labels)
-        parsed_data = []
-        max_index = 0
+        xyz_dict = {"X": 0, "Y": 1, "Z": 2}
+
+        # 3-dimensional ndarray (XYZ, terms, register)
+        self._spin_array = np.zeros((3, len(labels), self._register_length), dtype=np.uint8)
         for term, label in enumerate(labels):
-            label_list = label.split()
-            for single_label in label_list:
-                xyz, nums = single_label.split("_", 1)
-                index_str, power_str = nums.split("^", 1) if "^" in nums else (nums, "1")
-
-                index = int(index_str)
-                power = int(power_str)
-                max_index = max(max_index, index)
-
-                if xyz in self._XYZ_DICT:
-                    parsed_data.append((term, self._XYZ_DICT[xyz], index, power))
-
-        self._register_length = max_index + 1
-        self._spin_array = np.zeros((3, num_terms, self._register_length), dtype=np.uint8)
-        for term, xyz_num, index, power in parsed_data:
-            register = self._register_length - index - 1
-
-            # Check the order of X, Y, and Z whether it has been already assigned.
-            if self._spin_array[range(xyz_num + 1, 3), term, register].any():
-                raise ValueError("Label must be XYZ order.")
-            # same label is not assigned.
-            if self._spin_array[xyz_num, term, register]:
-                raise ValueError("Duplicate label.")
-
-            self._spin_array[xyz_num, term, register] = power
+            for splitted_label in label.split():
+                xyz, nums = splitted_label.split("_", 1)
+
+                if xyz not in xyz_dict:
+                    continue
+
+                xyz_num = xyz_dict[xyz]
+                index, power = map(int, nums.split("^", 1)) if "^" in nums else (int(nums), 1)
+                if index >= self._register_length:
+                    raise ValueError("index must be smaller than register_length.")
```

Comment:
Its nice/helpful to have the value(s) being referred to be in the message as well ie index and register_length, though it might be more meaningful to a end user to have the whole label there so they can see it in context better